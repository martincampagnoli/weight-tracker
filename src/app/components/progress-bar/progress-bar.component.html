<div class="w-100">
  @if (progress()! < 100) {
  <span class="goal-label">to reach the goal: {{ goal$ | async }} kgs</span>
  } @else {
  <span class="goal-label"
    >Congratulations! Goal of {{ goal$ | async }} kgs reached!</span
  >
  }
  <mat-progress-bar
    class="goal-progress-bar"
    [mode]="'buffer'"
    [value]="progress()"
    [bufferValue]="0"
  >
  </mat-progress-bar>
</div>
