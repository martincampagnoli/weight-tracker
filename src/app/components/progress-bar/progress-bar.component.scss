.goal-label {
  font-style: italic;
}

.goal-progress-bar {
  width: 100%;
  min-width: 80px;
}
