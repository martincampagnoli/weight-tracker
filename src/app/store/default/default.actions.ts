import { createAction, props } from '@ngrx/store';
import { Entry } from 'src/app/models/Entry';

/**
 * Action representing a failure during data retrieval.
 */
export const getFailure = createAction(
  '[Default] GetFailure  Error',
  props<{ payload: { message: string } }>()
);

/**
 * Action to initiate retrieval of data.
 */
export const getData = createAction('[Entry] Get Entries');

/**
 * Action indicating successful retrieval of data.
 */
export const getDataSuccess = createAction(
  '[Entry] Get Entries Success',
  props<{ payload: Entry[] }>()
);

/**
 * Action to add a new entry.
 */
export const addEntrySuccess = createAction(
  '[Default] Add Entry Success',
  props<{ payload: { entry: Entry } }>()
);

/**
 * Action to delete a new entry.
 */
export const deleteEntrySuccess = createAction(
  '[Default] Delete Entry Success',
  props<{ payload: { id: number } }>()
);

/**
 * Action to reset the application state.
 */
export const resetAppState = createAction('[Default] Reset app state');
