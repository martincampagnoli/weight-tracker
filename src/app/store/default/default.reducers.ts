import {
  Action,
  createFeatureSelector,
  createReducer,
  createSelector,
  on,
} from '@ngrx/store';
import {
  createEntry,
  deleteEntry,
  getDataSuccess,
  resetAppState,
  resetGoal,
  setGoal,
} from './default.actions';
import { Entry } from 'src/app/models/Entry';

/**
 * The structure of the application state.
 */
export interface AppState {
  loading: boolean;
  data: Entry[];
  nextId: number;
  targetWeight?: number;
}

/**
 * The default initial state for the application.
 */
const initialState: AppState = {
  loading: false,
  data: [],
  nextId: 1,
  targetWeight: undefined,
};

/**
 * Reducer function that handles state changes based on dispatched actions.
 *
 * @param state The current state.
 * @param action The dispatched action.
 * @returns The new state after applying the action.
 */
const reducer = createReducer(
  initialState,
  on(resetAppState, (initialState) => ({
    ...initialState,
  })),
  on(getDataSuccess, (state, { payload }) => {
    const maxId =
      payload.length > 0 ? Math.max(...payload.map((entry) => entry.id)) : 0;
    return {
      ...state,
      loading: false,
      data: payload,
      nextId: maxId + 1,
    };
  }),
  on(deleteEntry, (state, { payload }) => {
    return {
      ...state,
      loading: false,
      data: state.data.filter((entry) => entry.id !== payload.id),
    };
  }),
  on(createEntry, (state, { payload }) => {
    const newEntry: Entry = {
      id: state.nextId,
      weight: payload.weight,
      date: payload.date,
      description: payload.description,
    };
    return {
      ...state,
      loading: false,
      data: [newEntry, ...state.data],
      nextId: state.nextId + 1,
    };
  }),
  on(setGoal, (state, { payload }) => {
    const newGoal = payload.targetWeight;
    return {
      ...state,
      targetWeight: newGoal,
    };
  }),
  on(resetGoal, (state) => {
    return {
      ...state,
      targetWeight: undefined,
    };
  })
);

/**
 * The reducer function that handles various app-related actions to modify the state.
 *
 * @param state The current state.
 * @param action The dispatched action.
 * @returns The new state after applying the action.
 */
export function appReducer(state: AppState | undefined, action: Action) {
  return reducer(state, action);
}

/**
 * Selects the 'app' feature state.
 */
export const selectAppState = createFeatureSelector<AppState>('app');

/**
 * Selectors for various parts of the application state.
 */
export const getAppState = createSelector(
  selectAppState,
  (state: AppState) => state
);

export const getDataState = createSelector(
  selectAppState,
  (state: AppState) => state.data
);

/**
 * Selector to get the next available ID for new entries.
 */
export const getNextEntryId = createSelector(
  selectAppState,
  (state: AppState): number => state.nextId
);

/**
 * Selector to get the target weight.
 */
export const getTargetWeight = createSelector(
  selectAppState,
  (state: AppState) => state.targetWeight
);
