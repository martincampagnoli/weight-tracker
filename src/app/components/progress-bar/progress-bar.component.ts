import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { MatProgressBar } from '@angular/material/progress-bar';
import { Store } from '@ngrx/store';
import { combineLatest, map } from 'rxjs';
import { GoalService } from 'src/app/services/goal.service';
import * as reducers from 'src/app/store/default';

/**
 * Component representing a progress bar.
 */
@Component({
  selector: 'app-progress-bar',
  templateUrl: './progress-bar.component.html',
  styleUrls: ['./progress-bar.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [MatProgressBar, CommonModule],
  standalone: true,
  host: { class: 'w-75' },
})
export class ProgressBarComponent {
  private goalService = inject(GoalService);
  store = inject(Store);

  private entries$ = this.store.select(reducers.getDataState);
  protected goal$ = this.store.select(reducers.getTargetWeight);

  protected progress = toSignal(
    combineLatest([this.entries$, this.goal$]).pipe(
      map(([entries, goal]) => {
        if (!entries || !goal) return 0;
        return this.goalService.calculateGoalProgress(entries, goal);
      })
    )
  );
}
