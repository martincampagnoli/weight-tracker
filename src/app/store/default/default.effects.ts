import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { of, catchError, switchMap } from 'rxjs';
import { DataService } from 'src/app/services/data.service';
import { getFailure, getData } from './default.actions';
import { getDataSuccess } from './default.actions';
import { Entry } from 'src/app/models/Entry';

/**
 * Effects for managing application actions related to data.
 */
@Injectable()
export class AppEffects {
  /**
   * Constructs the AppEffects.
   *
   * @param actions$ Observable stream of dispatched actions.
   * @param moviesService The service for fetching data.
   */
  constructor(private actions$: Actions, private dataService: DataService) {}

  /**
   * Effect to fetch data.
   */
  getData = createEffect(() =>
    this.actions$.pipe(
      ofType(getData),
      switchMap(() =>
        this.dataService.getData().pipe(
          switchMap((data: Entry[]) => [getDataSuccess({ payload: data })]),
          catchError((error) => of(getFailure({ payload: error })))
        )
      )
    )
  );
}
