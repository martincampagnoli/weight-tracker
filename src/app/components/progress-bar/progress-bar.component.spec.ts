import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProgressBarComponent } from './progress-bar.component';
import { GoalService } from 'src/app/services/goal.service';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import * as reducers from 'src/app/store/default';

describe('ProgressBarComponent', () => {
  let component: ProgressBarComponent;
  let fixture: ComponentFixture<ProgressBarComponent>;
  let element: HTMLElement;
  let store: MockStore;

  const mockGoalService = {
    calculateGoalProgress: jest.fn().mockReturnValue(75),
  };

  const initialState = {} as any;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ProgressBarComponent],
      providers: [
        { provide: GoalService, useValue: mockGoalService },
        provideMockStore({ initialState }),
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    store = TestBed.inject(MockStore);
    store.overrideSelector(reducers.getDataState, []);
    store.overrideSelector(reducers.getTargetWeight, 75);
    fixture = TestBed.createComponent(ProgressBarComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
