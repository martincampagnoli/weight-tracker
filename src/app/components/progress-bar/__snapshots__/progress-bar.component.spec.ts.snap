// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarComponent should match snapshot 1`] = `
<app-progress-bar
  entries$={[Function _Store]}
  goal$={[Function _Store]}
  goalService={[Function Object]}
  progress={[Function Function]}
  store={[Function _MockStore]}
>
  <div
    class="w-100"
  >
    <span
      class="goal-label"
    >
      to reach the goal: 75 kgs
    </span>
    <mat-progress-bar
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="75"
      class="mat-mdc-progress-bar mdc-linear-progress goal-progress-bar mat-primary mdc-linear-progress--animation-ready"
      mode="buffer"
      role="progressbar"
      tabindex="-1"
    >
      <div
        aria-hidden="true"
        class="mdc-linear-progress__buffer"
      >
        <div
          class="mdc-linear-progress__buffer-bar"
          style="flex-basis: 0%;"
        />
        <div
          class="mdc-linear-progress__buffer-dots"
        />
      </div>
      <div
        aria-hidden="true"
        class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
        style="transform: scaleX(0.75);"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
      <div
        aria-hidden="true"
        class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
    </mat-progress-bar>
  </div>
</app-progress-bar>
`;
